# Notebook 01 — Inspección Inicial
**Proyecto Integrador | Minería de Datos I**

## Objetivo
Realizar una primera exploración del dataset crudo `streaming_users_dirty` para entender su estructura, detectar problemas de calidad y definir qué necesita limpiarse antes del análisis.

## Etapa
Esta es la etapa 1 de 5. No se modifica ningún dato, solo se observa y documenta.

In [ ]:
# Paso 1:  

import pandas as pd

# Leer el archivo JSON usando ruta relativa (sube un nivel con '..')
df = pd.read_json("../data/raw/streaming_users_dirty.json")

In [2]:
# Paso 2: Las primeras 5 columnas
df.head()

,user_id,age,subscription_plan,monthly_watch_time_mins,country,favorite_genre,last_login_date,customer_support_tickets
0,10000,39,Estándar,805.8,Brasil,Crime,2025-03-04,99
1,10001,37,Estándar,1173.4,Colombia,Crime,2019-04-02,2
2,10002,28,Básico,401.0,Colombia,Crime,2018-04-13,0
3,10003,43,Básico,62.4,Uruguay,Thriller,2021-01-31,0
4,10004,51,Básico,477.8,Perú,Thriller,2020-09-30,1


In [3]:
# Últimas Filas del Dataset
df.tail()

,user_id,age,subscription_plan,monthly_watch_time_mins,country,favorite_genre,last_login_date,customer_support_tickets
8155,10923,23,Premium,1161.4,Colombia,Romance,2023-05-15,0
8156,16525,27,Básico,436.2,Uruguay,Documental,2018-09-06,4
8157,11222,13,Estándar,1321.8,México,Documental,2019-02-08,0
8158,15613,38,Estándar,835.7,Brasil,Drama,2022-02-05,0
8159,16912,25,Estándar,1468.7,Argentina,Romance,2022-08-12,3


In [4]:
# Paso 3: Averiguo las dimensiones del dataset
df.shape


(8160, 8)

In [5]:
# Paso 4: 
# Inspeccionar los tipos de datos de cada columna y nulos generales
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8160 entries, 0 to 8159
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   user_id                   8160 non-null   int64  
 1   age                       8160 non-null   int64  
 2   subscription_plan         8160 non-null   str    
 3   monthly_watch_time_mins   7967 non-null   float64
 4   country                   8160 non-null   str    
 5   favorite_genre            7920 non-null   str    
 6   last_login_date           7840 non-null   str    
 7   customer_support_tickets  8160 non-null   int64  
dtypes: float64(1), int64(3), str(4)
memory usage: 756.7 KB


In [6]:
# Paso 5
# Contar cuántos valores nulos (faltantes) tiene cada columna
df.isnull().sum()

user_id                       0
age                           0
subscription_plan             0
monthly_watch_time_mins     193
country                       0
favorite_genre              240
last_login_date             320
customer_support_tickets      0
dtype: int64

In [7]:
# Paso 6:

# Contar la cantidad total de filas duplicadas en el dataset
df.duplicated().sum()

np.int64(126)

In [8]:
# Paso 7: Hacemos la Estadistica Descriptiva para ver los rangos minimo, maximo, cuartiles, media, desviacion estandar
df.describe()


,user_id,age,monthly_watch_time_mins,customer_support_tickets
count,8160.000000,8160.000000,7967.000000,8160.000000
mean,13995.433824,34.096814,1107.346153,1.800980
std,2310.810660,14.511304,5310.442604,11.334969
min,10000.000000,-5.000000,-120.000000,-1.000000
25%,11987.750000,25.000000,489.200000,0.000000
50%,13998.500000,33.000000,757.400000,1.000000
75%,15997.250000,42.000000,1045.700000,1.000000
max,17999.000000,150.000000,99999.000000,150.000000


In [9]:
# Observaciones
# Outliers evidentes
print("Edades negativas o cero:", (df['age'] <= 0).sum())
print("Edades > 100:", (df['age'] > 100).sum())
print("Tiempo negativo:", (df['monthly_watch_time_mins'] < 0).sum())
print("Tiempo > 10000:", (df['monthly_watch_time_mins'] > 10000).sum())
print("Tickets negativos:", (df['customer_support_tickets'] < 0).sum())

Edades negativas o cero: 45
Edades > 100: 53
Tiempo negativo: 49
Tiempo > 10000: 31
Tickets negativos: 29


In [10]:
# Paso 8: Valores únicos por columna categórica
for col in ['subscription_plan', 'country', 'favorite_genre']:
    print(f"{col} ({df[col].nunique()} valores únicos):")
    print(df[col].unique())
    print()

subscription_plan (15 valores únicos):
<ArrowStringArray>
[ 'Estándar',    'Básico',   'Premium',       'Std',  'estandar',    'basico',
    'básico',  'Premium ',   'premium',   'Premiun',    'BASICO',  'STANDARD',
     'Basic', 'Estándar ',   'PREMIUM']
Length: 15, dtype: str

country (26 valores únicos):
<ArrowStringArray>
[    'Brasil',   'Colombia',    'Uruguay',       'Perú',      'Chile',
  'Argentina',     'México',     'Brazil',     'brasil',     'méxico',
      'chile',    'uruguay',        'MEX',        'ARG',   'colombia',
        'COL',     'Mexico',        'URY',     'Chile ',       'Peru',
       'perú',  'argentina',        'PER', 'Argentina ',        'BRA',
        'CHL']
Length: 26, dtype: str

favorite_genre (28 valores únicos):
<ArrowStringArray>
[      'Crime',    'Thriller',       'Drama',      'Acción',     'Romance',
     'Comedia',  'Documental',      'ACCIÓN',           nan,       'CRIME',
    'Comedia ',      'comedy',       'DRAMA',    'THRILLER', 'Documenta

In [17]:
# Ejemplo de inconsistencia en subscription_plan
print("Variantes de 'Básico':", df[df['subscription_plan'].str.lower().str.contains('basic|básico')]['subscription_plan'].unique())

# Ejemplo de inconsistencia en country
print("Variantes de country:", df['country'].unique())

# Ejemplo de inconsistencia en favorite_genre
print("Variantes de favorite_genre:", df['favorite_genre'].unique())

Variantes de 'Básico': <ArrowStringArray>
['Básico', 'basico', 'básico', 'BASICO', 'Basic']
Length: 5, dtype: str
Variantes de country: <ArrowStringArray>
[    'Brasil',   'Colombia',    'Uruguay',       'Perú',      'Chile',
  'Argentina',     'México',     'Brazil',     'brasil',     'méxico',
      'chile',    'uruguay',        'MEX',        'ARG',   'colombia',
        'COL',     'Mexico',        'URY',     'Chile ',       'Peru',
       'perú',  'argentina',        'PER', 'Argentina ',        'BRA',
        'CHL']
Length: 26, dtype: str
Variantes de favorite_genre: <ArrowStringArray>
[      'Crime',    'Thriller',       'Drama',      'Acción',     'Romance',
     'Comedia',  'Documental',      'ACCIÓN',           nan,       'CRIME',
    'Comedia ',      'comedy',       'DRAMA',    'THRILLER', 'Documentary',
   'Thriller ',      'accion',      'Action',     'COMEDIA',     'ROMANCE',
     'thriler',         'DOC',     'romance',      'Crimen',      'Drama ',
       'drama',    'Roma

In [13]:
# Paso 9: Porcentaje de nulos por columna
(df.isnull().sum() / len(df) * 100).round(2)

user_id                     0.00
age                         0.00
subscription_plan           0.00
monthly_watch_time_mins     2.37
country                     0.00
favorite_genre              2.94
last_login_date             3.92
customer_support_tickets    0.00
dtype: float64

In [14]:
# Paso 10: Duplicados por user_id
print(f"User_ids duplicados: {df['user_id'].duplicated().sum()}")
print(f"User_ids únicos: {df['user_id'].nunique()}")
# Ver ejemplos
df[df['user_id'].duplicated(keep=False)].sort_values('user_id').head(50)

User_ids duplicados: 160
User_ids únicos: 8000


,user_id,age,subscription_plan,monthly_watch_time_mins,country,favorite_genre,last_login_date,customer_support_tickets
37,10037,33,Básico,611.0,Colombia,Documental,2019-09-29,2
8133,10037,33,Básico,611.0,Colombia,Documental,2019-09-29,2
52,10052,25,Básico,465.7,Colombia,Acción,2022-03-31,1
8089,10052,25,Básico,465.7,Colombia,Acción,2022-03-31,1
59,10059,39,Estándar,2976.6,Brasil,DRAMA,2024-06-22,1
8090,10059,39,Estándar,824.8,Brasil,Drama,2024-06-22,1
8010,10117,29,Estándar,713.9,Brasil,Documental,2020-12-20,1
117,10117,29,Estándar,713.9,Brasil,Documental,2020-12-20,1
128,10128,19,Básico,638.7,Argentina,Drama,2020-06-17,1
8085,10128,19,Básico,638.7,Argentina,Drama,2020-06-17,1


In [23]:
# Analizar las fechas rotas sin alterar el dataset original
print("--- ANÁLISIS DE CALIDAD EN FECHAS ---")

# Trabajamos sobre una copia temporal de la columna sin nulos para no romper nada
columna_fecha = df['last_login_date'].dropna()

# 1. Detectar textos basura que no son fechas
textos_basura = columna_fecha[columna_fecha.isin(['not_available', '0000-00-00'])]
print(f"Registros con texto inválido ('not_available'): {len(textos_basura)}")

# 2. Detectar fechas con problemas de lógica (meses mayores a 12 o días imposibles)
# Usamos to_datetime de forma aislada para ver qué falla, sin tocar el dataframe
fechas_limpias_temp = columna_fecha[~columna_fecha.isin(['not_available', '0000-00-00'])]
imposibles = fechas_limpias_temp[pd.to_datetime(fechas_limpias_temp, errors='coerce', format='%Y-%m-%d').isna()]
print(f"Registros con fechas lógicamente imposibles (ej. mes 15 o día 40): {len(imposibles)}")
if len(imposibles) > 0:
    print(f"Ejemplos de fechas imposibles: {imposibles.head(5).values}")

# 3. Detectar fechas que viajan al futuro (posteriores a hoy)
# Primero convertimos temporalmente lo que sí se puede para comparar
fechas_validas_convertidas = pd.to_datetime(fechas_limpias_temp, errors='coerce', format='%Y-%m-%d')
futuras = fechas_validas_convertidas[fechas_validas_convertidas > '2026-06-27']
print(f"\nRegistros con fechas en el futuro (posteriores a junio 2026): {len(futuras)}")
if len(futuras) > 0:
    # Mostramos cómo se ven originalmente en el formato sucio
    print(f"Ejemplos de fechas futuras detectadas: {df.loc[futuras.index, 'last_login_date'].head(5).values}")

--- ANÁLISIS DE CALIDAD EN FECHAS ---
Registros con texto inválido ('not_available'): 31
Registros con fechas lógicamente imposibles (ej. mes 15 o día 40): 418
Ejemplos de fechas imposibles: <ArrowStringArray>
['11-19-2018', '01-06-2025', '2026-15-03', '11-14-2025', '03-16-2024']
Length: 5, dtype: str

Registros con fechas en el futuro (posteriores a junio 2026): 15
Ejemplos de fechas futuras detectadas: <ArrowStringArray>
['2029-01-01', '2029-01-01', '2029-01-01', '2029-01-01', '2029-01-01']
Length: 5, dtype: str


### Lo que encontré al revisar los datos en bruto

El dataset llegó con varios problemas de calidad que voy a tener que corregir antes de cualquier análisis. Tiene 8160 filas y 8 columnas, mezclando datos personales con métricas de consumo.

**Valores faltantes:** No son una locura, pero hay que atenderlos. Me faltan 193 tiempos de visualización, 240 géneros favoritos y 320 fechas de último login. El resto está completo.

**Duplicados:** Hay 126 filas calcadas que ensucian los números. Pero el problema más feo son los 160 `user_id` repetidos (80 usuarios con dos registros). Algunos tienen datos distintos, así que no sé si son actualizaciones mal cargadas o errores de base. Voy a tener que definir un criterio para decidir con cuál quedarme.

**Categorías inconsistentes:** Un quilombo. `subscription_plan` tiene 15 variantes para 3 planes que debería tener. `country` aparece de 26 formas distintas (Brasil, Brazil, brasil, BRA...). Y `favorite_genre` mezcla inglés, español, mayúsculas y errores ortográficos. Todo va a tener que unificarse.

**Outliers:** Hay cosas que no cierran. Edades negativas y gente de más de 100 años. Tiempos de visualización en negativo y otros que superan los 10.000 minutos (imposible en un mes). Tickets de soporte negativos. Todo eso hay que revisarlo.

**Fechas rotas:** `last_login_date` es texto, no fecha. Tiene `not_available`, `0000-00-00`, fechas en 2029, meses 15 y formatos mezclados (YYYY-MM-DD, DD-MM-YYYY, etc.). Un dolor de cabeza que voy a tener que normalizar.

**El plan:** Limpiar duplicados, unificar categorías, corregir outliers, arreglar las fechas y decidir qué hacer con los nulos. Recién después de eso voy a poder analizar en serio.

Preguntas que me surgen para el análisis (EDA):
Mirando estas variables, me interesaría responder más adelante cosas como:

1- ¿El plan de suscripción que elige la gente cambia según el país o la edad, o es parecido en todos lados?

2- ¿Los usuarios que más minutos consumen al mes son los mismos que terminan abriendo más tickets de soporte técnico?

3- ¿Hay algún género de contenido que predomine según el rango de edad de los usuarios?

4- ¿Los valores nulos en el tiempo de visualización se concentran en algún plan de suscripción en particular o pasará en cualquiera?